back/app/services/projects_finder_admin_service.rb

```diff
@@ -236,4 +238,42 @@ def self.parse_date(date_input)
       nil
     end
   end
+
+  # Filter projects by visibility (access rights)
+  def self.filter_visibility(scope, params = {})
+    visibility_params = Array(params[:visibility])
+    return scope if visibility_params.blank?
+
+    valid_visibilities = %w[public groups admins]
+    selected_visibilities = visibility_params & valid_visibilities
+
+    return scope if selected_visibilities.blank?
+
+    scope.where(visible_to: selected_visibilities)
+  end
+
+  # Filter projects by discoverability (listed/unlisted status)
+  def self.filter_discoverability(scope, params = {})
+    discoverability_params = Array(params[:discoverability])
+    return scope if discoverability_params.blank?
+
+    valid_discoverabilities = %w[listed unlisted]
+    selected_discoverabilities = discoverability_params & valid_discoverabilities
+
+    return scope if selected_discoverabilities.blank?
+
+    conditions = []
+
+    if selected_discoverabilities.include?('listed')
+      conditions << scope.klass.where(listed: true)
+    end
+
+    if selected_discoverabilities.include?('unlisted')
+      conditions << scope.klass.where(listed: false)
+    end
+
+    return scope if conditions.empty?
+
+    scope.merge(conditions.reduce(:or))
```

Comment:
```suggestion
    return scope if selected_discoverabilities.blank? || selected_discoverabilities.length == 2
    
    if selected_discoverabilities.include?('listed')
      return scope.where(listed: true)
    else
      return scope.where(listed:false)
    end
```

This already simplifies the logic a bit.

It could be even cleaner by modeling this as three values: `'listed'`, `'unlisted'`, and `nil`. There are only three ways this filter makes sense to be used: give me listed only, give me unlisted only, or give me everything. The way you modeled it, it seems like there are 4 options: give me listed only (`['listed']`), give me unlisted only (`['unlisted']`), give me everything (`['listed', 'unlisted']`), or give me nothing (`[]`). But because 'give me nothing' makes no sense, we have to add an extra condition to treat it as 'give me everything'. Anyway, not a huge problem and this way is also not bad, so up to you